### PR TITLE
Powered rails can kick-start minecarts

### LINKED ISSUE #3017: Powered rails can't kickstart a Minecart properly
You push the button and the cart should normally start - with or without a person riding it.
For some reason, it doesn't, for me. The cart instead seems to be pulled towards the stone block. A few times while testing I was partially clipped inside - and could go right through it if I used the walk buttons while sitting in the cart. It went in the wrong direction.

### PR DESCRIPTION
**Pull request by marvinkopf**
fixes #3017

### CODE REVIEW

File: src/Entities/Minecart.cpp [hunk 1]
```diff
@@ -698,6 +728,20 @@ void cMinecart::SnapToRail(NIBBLETYPE a_RailMeta)
 
 
 
+bool cMinecart::TestBlockCollision(int a_XOffset, int a_YOffset, int a_ZOffset)
+{
+	BLOCKTYPE Block = m_World->GetBlock(POSX_TOINT + a_XOffset, POSY_TOINT + a_YOffset, POSZ_TOINT + a_ZOffset);
```
Comment (sphinxc0re) [inline src/Entities/Minecart.cpp]: Where are these `POS#_TOINT` variables coming from?

File: src/Entities/Minecart.cpp [hunk 2]
```diff
@@ -777,13 +817,10 @@ bool cMinecart::TestBlockCollision(NIBBLETYPE a_RailMeta)
 		}
 		case E_META_RAIL_CURVED_ZM_XM:
 		{
-			BLOCKTYPE BlockXM = m_World->GetBlock(POSX_TOINT - 1, POSY_TOINT, POSZ_TOINT);
-			BLOCKTYPE BlockZM = m_World->GetBlock(POSX_TOINT, POSY_TOINT, POSZ_TOINT - 1);
+			bool BlockXM = TestBlockCollision(-1, 0, 0);
+			bool BlockZM = TestBlockCollision(0, 0, -1);
```
Comment (sphinxc0re) [inline src/Entities/Minecart.cpp]: Wouldn't it make sense to rename these since the type has changed?

File: src/Entities/Minecart.cpp [hunk 3]
```diff
@@ -794,13 +831,10 @@ bool cMinecart::TestBlockCollision(NIBBLETYPE a_RailMeta)
 		}
 		case E_META_RAIL_CURVED_ZM_XP:
 		{
-			BLOCKTYPE BlockXP = m_World->GetBlock(POSX_TOINT + 1, POSY_TOINT, POSZ_TOINT);
-			BLOCKTYPE BlockZM = m_World->GetBlock(POSX_TOINT, POSY_TOINT, POSZ_TOINT - 1);
+			bool BlockXP = TestBlockCollision(1, 0, 0);
+			bool BlockZM = TestBlockCollision(0, 0, -1);
```
Comment (sphinxc0re) [inline src/Entities/Minecart.cpp]: Wouldn't it make sense to rename these since the type has changed?

File: src/Entities/Minecart.cpp [hunk 4]
```diff
@@ -811,13 +845,10 @@ bool cMinecart::TestBlockCollision(NIBBLETYPE a_RailMeta)
 		}
 		case E_META_RAIL_CURVED_ZP_XM:
 		{
-			BLOCKTYPE BlockXM = m_World->GetBlock(POSX_TOINT - 1, POSY_TOINT, POSZ_TOINT);
-			BLOCKTYPE BlockZP = m_World->GetBlock(POSX_TOINT, POSY_TOINT, POSZ_TOINT + 1);
+			bool BlockXM = TestBlockCollision(-1, 0, 0);
+			bool BlockZP = TestBlockCollision(0, 0, +1);
```
Comment (sphinxc0re) [inline src/Entities/Minecart.cpp]: Wouldn't it make sense to rename these since the type has changed?

File: src/Entities/Minecart.cpp [hunk 5]
```diff
@@ -828,13 +859,10 @@ bool cMinecart::TestBlockCollision(NIBBLETYPE a_RailMeta)
 		}
 		case E_META_RAIL_CURVED_ZP_XP:
 		{
-			BLOCKTYPE BlockXP = m_World->GetBlock(POSX_TOINT + 1, POSY_TOINT, POSZ_TOINT);
-			BLOCKTYPE BlockZP = m_World->GetBlock(POSX_TOINT, POSY_TOINT, POSZ_TOINT + 1);
+			bool BlockXP = TestBlockCollision(1, 0, 0);
+			bool BlockZP = TestBlockCollision(0, 0, 1);
```
Comment (sphinxc0re) [inline src/Entities/Minecart.cpp]: Wouldn't it make sense to rename these since the type has changed?

File: src/Entities/Minecart.cpp [hunk 6]
```diff
@@ -698,6 +728,20 @@ void cMinecart::SnapToRail(NIBBLETYPE a_RailMeta)
 
 
 
+bool cMinecart::TestBlockCollision(int a_XOffset, int a_YOffset, int a_ZOffset)
```
Comment (sphinxc0re) [inline src/Entities/Minecart.cpp]: Have you thought about inlining this method?

Comment (marvinkopf) [inline src/Entities/Minecart.cpp]: For the performance? Usually compilers already do a good job deciding if a function should be inlined.

File: src/Entities/Minecart.cpp (hunk 1 above)
Comment (marvinkopf) [inline src/Entities/Minecart.cpp]: The macros are defined in the entity class.

File: src/Entities/Minecart.cpp (hunk 2 above)
Comment (marvinkopf) [inline src/Entities/Minecart.cpp]: Sure.

File: src/Entities/Minecart.cpp (hunk 6 above)
Comment (sphinxc0re) [inline src/Entities/Minecart.cpp]: Yeah, it should be alright 👍

File: src/Entities/Minecart.h [hunk 7]
```diff
@@ -78,6 +78,8 @@ class cMinecart :
 	void SnapToRail(NIBBLETYPE a_RailMeta);
 	/** Tests if a solid block is in front of a cart, and stops the cart (and returns true) if so; returns false if no obstruction */
 	bool TestBlockCollision(NIBBLETYPE a_RailMeta);
+	/** Tests if a solid block is at a specific offset of the minecart position */
+	bool IsBlockAtOffset(int a_XOffset, int a_YOffset, int a_ZOffset);
```
Comment (madmaxoft) [inline src/Entities/Minecart.h]: Please rename to `IsSolidBlockAtOffset` to make it clear what the function does.

**Comment by madmaxoft** [comment]
I thought I found a bug in the code, until I had a second look, and only then realized that it's not possible to trigger the behavior I thought was wrong. But it lead me to another realization - your code uses `north`, `south` etc., while we prefer to use `ZM`, `ZP` etc. Could you still change that please? I believe it's far easier to refer to the axis directly, than having to visualise which way "north" is in Minecraft axis system.

**Comment by sphinxc0re** [comment]
@madmaxoft why is it called `ZM` again and not `ZN`? That would make much more sense because it would resemble `z-negative`

**Comment by madmaxoft** [comment]
`Z Plus` and `Z Minus`

**Comment by marvinkopf** [comment]
It's alright, I've changed it.

**Pull request merged**